### PR TITLE
NET 6409

### PR DESCRIPTION
**Pull request by absolutelightning**
Changes proposed in this PR:

Fixes - https://github.com/hashicorp/consul-k8s/issues/2433

consul-k8s CLI: Display envoy cluster stats using consul-k8s CLI 

How I've tested this PR:

* Setup Kind cluster using [tutorial](https://github.com/hashicorp-education/learn-consul-get-started-kubernetes)
* `make cli-dev`
* `./consul-k8s proxy stats -pod nginx-6d7469694f-rdmwp -kubeconfig /Users/asheshvidyut/.kube/config -namespace default > nginx`
* `cat nginx`

```
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.version_text: ""
cluster.payments.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.version_text: ""
cluster.product-api-db.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.version_text: ""
cluster.product-api.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.version_text: ""
cluster.public-api.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.version_text: ""
cluster_manager.cds.version_text: ""
control_plane.identifier: ""
listener_manager.lds.version_text: ""
cluster.consul-dataplane.assignment_stale: 0
cluster.consul-dataplane.assignment_timeout_received: 0
cluster.consul-dataplane.bind_errors: 0
cluster.consul-dataplane.circuit_breakers.default.cx_open: 0
cluster.consul-dataplane.circuit_breakers.default.cx_pool_open: 0
cluster.consul-dataplane.circuit_breakers.default.rq_open: 0
cluster.consul-dataplane.circuit_breakers.default.rq_pending_open: 0
cluster.consul-dataplane.circuit_breakers.default.rq_retry_open: 0
cluster.consul-dataplane.circuit_breakers.high.cx_open: 0
cluster.consul-dataplane.circuit_breakers.high.cx_pool_open: 0
cluster.consul-dataplane.circuit_breakers.high.rq_open: 0
cluster.consul-dataplane.circuit_breakers.high.rq_pending_open: 0
cluster.consul-dataplane.circuit_breakers.high.rq_retry_open: 0
cluster.consul-dataplane.default.total_match_count: 1
cluster.consul-dataplane.http2.deferred_stream_close: 0
cluster.consul-dataplane.http2.dropped_headers_with_underscores: 0
cluster.consul-dataplane.http2.header_overflow: 0
cluster.consul-dataplane.http2.headers_cb_no_stream: 0
cluster.consul-dataplane.http2.inbound_empty_frames_flood: 0
cluster.consul-dataplane.http2.inbound_priority_frames_flood: 0
cluster.consul-dataplane.http2.inbound_window_update_frames_flood: 0
cluster.consul-dataplane.http2.keepalive_timeout: 0
cluster.consul-dataplane.http2.metadata_empty_frames: 0
cluster.consul-dataplane.http2.outbound_control_flood: 0
cluster.consul-dataplane.http2.outbound_flood: 0
cluster.consul-dataplane.http2.pending_send_bytes: 0
cluster.consul-dataplane.http2.requests_rejected_with_underscores_in_headers: 0
cluster.consul-dataplane.http2.rx_messaging_error: 0
cluster.consul-dataplane.http2.rx_reset: 0
cluster.consul-dataplane.http2.stream_refused_errors: 0
cluster.consul-dataplane.http2.streams_active: 1
cluster.consul-dataplane.http2.trailers: 0
cluster.consul-dataplane.http2.tx_flush_timeout: 0
cluster.consul-dataplane.http2.tx_reset: 0
cluster.consul-dataplane.internal.upstream_rq_200: 1
cluster.consul-dataplane.internal.upstream_rq_2xx: 1
cluster.consul-dataplane.internal.upstream_rq_completed: 1
cluster.consul-dataplane.lb_healthy_panic: 0
cluster.consul-dataplane.lb_local_cluster_not_ok: 0
cluster.consul-dataplane.lb_recalculate_zone_structures: 0
cluster.consul-dataplane.lb_subsets_active: 0
cluster.consul-dataplane.lb_subsets_created: 0
cluster.consul-dataplane.lb_subsets_fallback: 0
cluster.consul-dataplane.lb_subsets_fallback_panic: 0
cluster.consul-dataplane.lb_subsets_removed: 0
cluster.consul-dataplane.lb_subsets_selected: 0
cluster.consul-dataplane.lb_zone_cluster_too_small: 0
cluster.consul-dataplane.lb_zone_no_capacity_left: 0
cluster.consul-dataplane.lb_zone_number_differs: 0
cluster.consul-dataplane.lb_zone_routing_all_directly: 0
cluster.consul-dataplane.lb_zone_routing_cross_zone: 0
cluster.consul-dataplane.lb_zone_routing_sampled: 0
cluster.consul-dataplane.max_host_weight: 0
cluster.consul-dataplane.membership_change: 1
cluster.consul-dataplane.membership_degraded: 0
cluster.consul-dataplane.membership_excluded: 0
cluster.consul-dataplane.membership_healthy: 1
cluster.consul-dataplane.membership_total: 1
cluster.consul-dataplane.original_dst_host_invalid: 0
cluster.consul-dataplane.retry_or_shadow_abandoned: 0
cluster.consul-dataplane.update_attempt: 0
cluster.consul-dataplane.update_empty: 0
cluster.consul-dataplane.update_failure: 0
cluster.consul-dataplane.update_no_rebuild: 0
cluster.consul-dataplane.update_success: 0
cluster.consul-dataplane.upstream_cx_active: 1
cluster.consul-dataplane.upstream_cx_close_notify: 0
cluster.consul-dataplane.upstream_cx_connect_attempts_exceeded: 0
cluster.consul-dataplane.upstream_cx_connect_fail: 0
cluster.consul-dataplane.upstream_cx_connect_timeout: 0
cluster.consul-dataplane.upstream_cx_connect_with_0_rtt: 0
cluster.consul-dataplane.upstream_cx_destroy: 0
cluster.consul-dataplane.upstream_cx_destroy_local: 0
cluster.consul-dataplane.upstream_cx_destroy_local_with_active_rq: 0
cluster.consul-dataplane.upstream_cx_destroy_remote: 0
cluster.consul-dataplane.upstream_cx_destroy_remote_with_active_rq: 0
cluster.consul-dataplane.upstream_cx_destroy_with_active_rq: 0
cluster.consul-dataplane.upstream_cx_http1_total: 0
cluster.consul-dataplane.upstream_cx_http2_total: 1
cluster.consul-dataplane.upstream_cx_http3_total: 0
cluster.consul-dataplane.upstream_cx_idle_timeout: 0
cluster.consul-dataplane.upstream_cx_max_duration_reached: 0
cluster.consul-dataplane.upstream_cx_max_requests: 0
cluster.consul-dataplane.upstream_cx_none_healthy: 0
cluster.consul-dataplane.upstream_cx_overflow: 0
cluster.consul-dataplane.upstream_cx_pool_overflow: 0
cluster.consul-dataplane.upstream_cx_protocol_error: 0
cluster.consul-dataplane.upstream_cx_rx_bytes_buffered: 17
cluster.consul-dataplane.upstream_cx_rx_bytes_total: 24037
cluster.consul-dataplane.upstream_cx_total: 1
cluster.consul-dataplane.upstream_cx_tx_bytes_buffered: 0
cluster.consul-dataplane.upstream_cx_tx_bytes_total: 367029
cluster.consul-dataplane.upstream_flow_control_backed_up_total: 0
cluster.consul-dataplane.upstream_flow_control_drained_total: 0
cluster.consul-dataplane.upstream_flow_control_paused_reading_total: 0
cluster.consul-dataplane.upstream_flow_control_resumed_reading_total: 0
cluster.consul-dataplane.upstream_http3_broken: 0
cluster.consul-dataplane.upstream_internal_redirect_failed_total: 0
cluster.consul-dataplane.upstream_internal_redirect_succeeded_total: 0
cluster.consul-dataplane.upstream_rq_0rtt: 0
cluster.consul-dataplane.upstream_rq_200: 1
cluster.consul-dataplane.upstream_rq_2xx: 1
cluster.consul-dataplane.upstream_rq_active: 1
cluster.consul-dataplane.upstream_rq_cancelled: 0
cluster.consul-dataplane.upstream_rq_completed: 1
cluster.consul-dataplane.upstream_rq_maintenance_mode: 0
cluster.consul-dataplane.upstream_rq_max_duration_reached: 0
cluster.consul-dataplane.upstream_rq_pending_active: 0
cluster.consul-dataplane.upstream_rq_pending_failure_eject: 0
cluster.consul-dataplane.upstream_rq_pending_overflow: 0
cluster.consul-dataplane.upstream_rq_pending_total: 1
cluster.consul-dataplane.upstream_rq_per_try_idle_timeout: 0
cluster.consul-dataplane.upstream_rq_per_try_timeout: 0
cluster.consul-dataplane.upstream_rq_retry: 0
cluster.consul-dataplane.upstream_rq_retry_backoff_exponential: 0
cluster.consul-dataplane.upstream_rq_retry_backoff_ratelimited: 0
cluster.consul-dataplane.upstream_rq_retry_limit_exceeded: 0
cluster.consul-dataplane.upstream_rq_retry_overflow: 0
cluster.consul-dataplane.upstream_rq_retry_success: 0
cluster.consul-dataplane.upstream_rq_rx_reset: 0
cluster.consul-dataplane.upstream_rq_timeout: 0
cluster.consul-dataplane.upstream_rq_total: 1
cluster.consul-dataplane.upstream_rq_tx_reset: 0
cluster.consul-dataplane.version: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.assignment_stale: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.assignment_timeout_received: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.bind_errors: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.default.cx_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.default.cx_pool_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.default.rq_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.default.rq_pending_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.default.rq_retry_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.high.cx_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.high.cx_pool_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.high.rq_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.high.rq_pending_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.circuit_breakers.high.rq_retry_open: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.client_ssl_socket_factory.downstream_context_secrets_not_ready: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.client_ssl_socket_factory.ssl_context_update_by_sds: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.client_ssl_socket_factory.upstream_context_secrets_not_ready: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.default.total_match_count: 2
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.init_fetch_timeout: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_healthy_panic: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_local_cluster_not_ok: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_recalculate_zone_structures: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_subsets_active: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_subsets_created: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_subsets_fallback: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_subsets_fallback_panic: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_subsets_removed: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_subsets_selected: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_zone_cluster_too_small: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_zone_no_capacity_left: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_zone_number_differs: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_zone_routing_all_directly: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_zone_routing_cross_zone: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.lb_zone_routing_sampled: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.max_host_weight: 1
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.membership_change: 1
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.membership_degraded: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.membership_excluded: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.membership_healthy: 1
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.membership_total: 1
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.original_dst_host_invalid: 0
cluster.frontend.default.dc1.internal.79a003be-075c-5b47-4b7e-1dea66df560f.consul.outlier_detection.ejections_active: 0 ...........
```

How I expect reviewers to test this PR:

Checklist:
- [x] Tests added
- [x] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: cli/cmd/envoy-stats/envoy_stats_test.go
Comment (absolutelightning) [inline cli/cmd/envoy-stats/envoy_stats_test.go:62]: Need some pointers how can I test more.

**Comment by david-yu** [comment]
Thanks @absolutelightning Will provide feedback later this week, also pinging SMEs for feedback.

File: cli/cmd/envoy-stats/envoy_stats_test.go
Comment (Ganeshrockz) [reply cli/cmd/envoy-stats/envoy_stats_test.go:62]: Here's how I would go about testing this.

1. Start a Http server in tests and register the `/stats` route handler to it
2. Assert the response from the test case

This should pretty much give you the needed coverage

**Review by Ganeshrockz — COMMENTED** [review]
Looks good in general.

File: cli/cmd/envoy-stats/envoy_stats_test.go
Comment (Ganeshrockz) [reply cli/cmd/envoy-stats/envoy_stats_test.go:62]: You could work around the port forwarding logic by creating a mockPortForwarder just for the tests. In such case, `GetLocalPort` would return back the port of the exposed Http server

Comment (absolutelightning) [reply cli/cmd/envoy-stats/envoy_stats_test.go:66]: thanks

**Review by Ganeshrockz — COMMENTED** [review]
Looks good. Deferring the final approval to the SMEs

File: cli/cmd/envoy-stats/envoy_stats.go
Comment (t-eckert) [inline cli/cmd/envoy-stats/envoy_stats.go:190]:
```suggestion
	return "Display Envoy stats for a proxy"
```

File: cli/cmd/envoy-stats/envoy_stats.go
Comment (t-eckert) [inline cli/cmd/envoy-stats/envoy_stats.go:98]: It's more common in the Kubernetes ecosystem to use the namespace provided in the Kubeconfig if none is explicityly provided. This is how `kubectl` does it and we followed that model with all of the existing commands. If a user provides a value for `namespace`, it should be used. If not, fall back to whatever the value is in the Kubeconfig.

You can follow the pattern here: https://github.com/hashicorp/consul-k8s/blob/e50c8e20f039f33e0261409da91870daafc28aaa/cli/cmd/proxy/read/command.go#L298

**Review by t-eckert — CHANGES_REQUESTED** [review]
Ashesh, excellent work. I think this will be very useful for many of our practitioners. I know I will use it for debugging.

I left some requested changes that bring this work in line with the existing patterns and practices that we have for this CLI. We want the user interface for the CLI to be as consistent as possible. 

Please reach out if you have any questions about my feedback. I would love to help you get this PR merged as I think the functionality is great!

**Comment by absolutelightning** [comment]
> Ashesh, excellent work. I think this will be very useful for many of our practitioners. I know I will use it for debugging.
> 
> I left some requested changes that bring this work in line with the existing patterns and practices that we have for this CLI. We want the user interface for the CLI to be as consistent as possible.
> 
> Please reach out if you have any questions about my feedback. I would love to help you get this PR merged as I think the functionality is great!

Thanks for review 🙏 I have updated the PR please check again.

File: cli/cmd/proxy/stats/command.go
Comment (t-eckert) [inline cli/cmd/proxy/stats/command.go:125]: This is not quite the namespace you want to grab. This will get the namespace that Consul was installed into. This is generally `consul`, but we advise against users deploying their applications in the `consul` namespace. 

Instead, you will want to grab the namespace currently associated with the users Kubeconfig.

```go
    if c.flagNamespace == "" {
		c.flagNamespace = settings.Namespace()
	}
```

**Review by t-eckert — APPROVED** [review]
Great! Thank you so much for adding this functionality. I added some comments on things that I would like you to change (they are small), but I went ahead and approved so that you can merge when you get the issues fixed up.

The most important comment that needs to be fixed is grabbing the correct namespace. You'll need to grab it from the Helm settings instead of the deployment configuration. This is because we want it to reflect the configuration in the users Kubectl. The namespace of the deployment is not that and is not generally where users will want to look for Envoy deployments (unless they are looking for gateway configurations).

**Comment by david-yu** [comment]
I think this is great, thank you for doing this. I can't tell from the PR, but does the CLI provide a new line right after the command is issued? I think that is necessary to at least provide some separation between CLI output and the CLI command itself.

To close this out we should also provide docs for this feature here https://developer.hashicorp.com/consul/docs/k8s/k8s-cli#proxy and add it after the proxy log command. We should be explicit to also note that this is cluster stats we are outputting as opposed to XDS Server stats: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/statistics  

There may be additional feedback in the future but I think this is a great start and will tremendously help troubleshooting.

**Review by david-yu — APPROVED** [review]

**Comment by absolutelightning** [comment]
> I think this is great, thank you for doing this. I can't tell from the PR, but does the CLI provide a new line right after the command is issued? I think that is necessary to at least provide some separation between CLI output and the CLI command itself.
> 
> To close this out we should also provide docs for this feature here https://developer.hashicorp.com/consul/docs/k8s/k8s-cli#proxy and add it after the proxy log command. We should be explicit to also note that this is cluster stats we are outputting as opposed to XDS Server stats: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/statistics
> 
> There may be additional feedback in the future but I think this is a great start and will tremendously help troubleshooting.

Yes there is a new line after the command. It looks like this 
```
asheshvidyut@asheshvidyut-H2GX766V9T ~/consul-k8s (NET-6409 %) » ./consul-k8s proxy stats public-api-567d949866-452xc -kubeconfig /Users/asheshvidyut/.kube/config -n default
cluster.frontend.default.dc1.internal.4b11bae3-b8ca-ee63-89bc-428cbfa6ef60.consul.version_text: ""
cluster.nginx.default.dc1.internal.4b11bae3-b8ca-ee63-89bc-428cbfa6ef60.consul.version_text: ""
cluster.payments.default.dc1.internal.4b11bae3-b8ca-ee63-89bc-428cbfa6ef60.consul.version_text: ""
cluster.product-api-db.default.dc1.internal.4b11bae3-b8ca-ee63-89bc-428cbfa6ef60.consul.version_text: ""
cluster.product-api.default.dc1.internal.4b11bae3-b8ca-ee63-89bc-428cbfa6ef60.consul.version_text: ""
cluster_manager.cds.version_text: ""
control_plane.identifier: ""
listener_manager.lds.version_text: ""
cluster.consul-dataplane.assignment_stale: 0 ....
```

Also I have created the PR in consul to update the docs. - https://github.com/hashicorp/consul/pull/19515
Page can be viewed here - https://consul-git-net-6409-hashicorp.vercel.app/consul/docs/k8s/k8s-cli#proxy-stats

**Comment by david-yu** [comment]
Could we backport to 1.3.x, 1.2.x and 1.1.x?

**Comment by david-yu** [comment]
This can be merged without waiting for patch releases, only the docs need to wait for the releases.

**Comment by absolutelightning** [comment]
@t-eckert - Not able to merge. Tests are failing which are not related to my change.

**Pull request merged**
